### PR TITLE
Why we rewrote Lua in JS

### PR DESCRIPTION
**Pull request by giann**
A post about a project I've been working on with daurnimator for the past months.

### CODE REVIEW

File: posts/posts_meta.lua [hunk 1]
```diff
@@ -133,6 +133,16 @@ local authors = {
 		linkedin = nil,
 		facebook = nil,
 	},
+	giann = {
+		name = 'Benoit Giannangeli',
+		picture = 'https://avatars2.githubusercontent.com/u/584398?v=4&s=460',
+		description = 'Front-End Web and iOS Developer, Game Developer, Musician, Lua enthusiast.',
+		website = 'nil',
```
Comment (daurnimator) [inline posts/posts_meta.lua]: the string nil?

File: posts/webdev/why-we-rewrote-lua-in-js.md [hunk 2]
```diff
@@ -0,0 +1,78 @@
...
+There is a latent desire out there to use something else than JS in the browser. Whether it is warranted or not is another topic in itself. But you don't have to search long to find [all kinds of projects](https://github.com/jashkenas/coffeescript/wiki/List-of-languages-that-compile-to-JS) aiming to bring another language in the browser.
+
+The arrival of Web Assembly reignited that desire for many of us. But wasm is a long way from being usable in production and still does not allow full interaction with all web apis.
+
+## Lua is a good fit for the browser
+
+Lua is a simple language with very few concepts to understand and a clear and readable syntax. You can be proficient with it in a matter of hours. Yet it provides very useful features. To name a few:
+- First class functions and closures
+- A versatile data structure: [the table](http://www.lua.org/pil/2.5.html)
+- Spread operator
```
Comment (mpeterv) [inline posts/webdev/why-we-rewrote-lua-in-js.md]: There is no spread operator in Lua, the closest equivalent is `table.unpack`.

File: posts/webdev/why-we-rewrote-lua-in-js.md [hunk 3]
```diff
@@ -0,0 +1,78 @@
...
+- Lexical scoping
+- Return or assign multiple values
+- Iterators
+- Coroutines (see below)
+
+It has a sane ceorcion system (I'm looking at you JS!). Nobody's adding X new concepts to it each version (still looking at you JS), which makes it a stable and reliable language.
```
Comment (mpeterv) [inline posts/webdev/why-we-rewrote-lua-in-js.md]: Typo: `ceorcion` -> `coercion`

File: posts/webdev/why-we-rewrote-lua-in-js.md [hunk 4]
```diff
@@ -0,0 +1,78 @@
...
+
+Lua is already successfully used server-side with the awesome [OpenResty](https://openresty.org/en/) which powers sites like [itch.io](https://itch.io/). There's even some great web frameworks like [lapis](http://leafo.net/lapis/).
+
+Lua can be found in a number of widely different contexts because of how easy it is to embed it. You can write a whole program with it or simply use it has a *glue* language thanks to the Lua C api.
```
Comment (mpeterv) [inline posts/webdev/why-we-rewrote-lua-in-js.md]: Typo: `use it has a` -> `use it as a`

File: posts/webdev/why-we-rewrote-lua-in-js.md [hunk 5]
```diff
@@ -0,0 +1,78 @@
...
+## Existing ways of using Lua in the browser
+
+There's already a few projects out there to use Lua in the browser in some extent:
+- [Moonshine](http://moonshinejs.org/) is a reimplementation of Lua in JS. Sadly, it's not being actively developed anymore.
+- [Starlight](http://starlight.paulcuth.me.uk/) by the same author is a Lua to JS transpiler. Unfortunately coroutines can't be implemented effectively with this approach.
+- [lua.vm.js](https://daurnimator.github.io/lua.vm.js/lua.vm.js.html) a port of the original Lua implementation to JS using Emscripten. It's fast and works well but its interoperability with JS is compromised by the fact that we end up with two garbage collectors trying to manage the same memory.
+
+## Fengari
+
+[**Fengari**](http://fengari.io/) (moon in greek) is the Lua VM written in Javascript with the browser as its primary target.
```
Comment (mpeterv) [inline posts/webdev/why-we-rewrote-lua-in-js.md]: It would be nice to explicitly mention that Fengari is based on Lua 5.3.

**Comment by mpeterv** [comment]
This is a nice introductory post, I'd love to read more. In particular, it would be great to have benchmarks against alternatives mentioned in the post. It would also be nice to read about what problems arose (why is compliance only at 99%) and what are advantages and disadvantages of JS as a host language as opposed to C.

**Comment by giann** [comment]
99% is just the current progress. The missing features are io functions that don't translate well in Node. But it's not available in the browser so really it's 100% if you're in the browser.

We'll certainly write a follow up on the _how_ later :)

**Comment by daurnimator** [comment]
There's also the missing lua_gc functionality

**Comment by giann** [comment]
Fixed typo and changed _spread operator_ to _vararg expression_

**Comment by Etiene** [comment]
Great post! it's published now 💃

**Pull request merged**